src/execution/operator/join/perfect_hash_join_executor.cpp

```diff
@@ -0,0 +1,221 @@
+}
+
+bool PerfectHashJoinExecutor::CanDoPerfectHashJoin() {
+	return pjoin_stats.is_build_small;
+}
+
+void PerfectHashJoinExecutor::BuildPerfectHashTable(JoinHashTable *hash_table_ptr, JoinHTScanState &join_ht_state,
+                                                    LogicalType key_type) {
+	// First, allocate memory for each build column
+	auto build_size = pjoin_stats.build_range + 1;
+	for (auto type : hash_table_ptr->build_types) {
+		perfect_hash_table.emplace_back(type, build_size);
+	}
+	// and for duplicate_checking
+	bitmap_build_idx = unique_ptr<bool[]>(new bool[build_size]);
+	memset(bitmap_build_idx.get(), 0, sizeof(bool) * build_size); // set false
+
+	// Now fill columns with build data
+	FullScanHashTable(join_ht_state, key_type, hash_table_ptr);
+}
+
+void PerfectHashJoinExecutor::FullScanHashTable(JoinHTScanState &state, LogicalType key_type,
+                                                JoinHashTable *hash_table) {
+	Vector tuples_addresses(LogicalType::POINTER, hash_table->size());      // allocate space for all the tuples
+	auto key_locations = FlatVector::GetData<data_ptr_t>(tuples_addresses); // get a pointer to vector data
+	// TODO: In a parallel finalize: One should exclusivly lock and each thread should do one part of the code below.
+	// Go through all the blocks and fill the keys addresses
+	auto keys_count = hash_table->FillWithHTOffsets(key_locations, state);
+	// Scan the build keys in the hash table
+	Vector build_vector(key_type, keys_count);
+	RowOperations::FullScanColumn(hash_table->layout, tuples_addresses, build_vector, keys_count, 0);
+	// Now fill the selection vector using the build keys and create a sequential vector
+	// todo: add check for fast pass when probe is part of build domain
+	SelectionVector sel_build(keys_count + 1);
+	SelectionVector sel_tuples(keys_count + 1);
+	FillSelectionVectorSwitchBuild(build_vector, sel_build, sel_tuples, keys_count);
+	// early out
+	if (has_duplicates)
+		return;
+	if (unique_keys == pjoin_stats.build_range + 1 && !hash_table->has_null) {
+		pjoin_stats.is_build_dense = true;
+	}
+	keys_count = unique_keys; // do not condider keys out of the range
+	// Full scan the remaining build columns and fill the perfect hash table
+	for (idx_t i = 0; i < hash_table->build_types.size(); i++) {
+		auto &vector = perfect_hash_table[i];
+		D_ASSERT(vector.GetType() == hash_table->build_types[i]);
+		const auto col_no = hash_table->condition_types.size() + i;
+		const auto col_offset = hash_table->layout.GetOffsets()[col_no];
+		RowOperations::Gather(tuples_addresses, sel_tuples, vector, sel_build, keys_count, col_offset, col_no);
+	}
+}
+
+void PerfectHashJoinExecutor::FillSelectionVectorSwitchBuild(Vector &source, SelectionVector &sel_vec,
+                                                             SelectionVector &seq_sel_vec, idx_t count) {
+	switch (source.GetType().id()) {
```

Comment:
Can we switch on physical type here? That could enable the build to work on dates/times/timestamps as well.